barretenberg/cpp/src/barretenberg/vm2/constraining/recursion/goblin_avm_recursive_verifier.hpp

```diff
@@ -79,47 +83,109 @@ class AvmGoblinRecursiveVerifier {
                                           const std::vector<std::vector<UltraFF>>& public_inputs,
                                           AggregationObject input_agg_obj) const
     {
-        using FF = AvmRecursiveFlavor::FF;
+        // Construct and prove the inner Mega-arithmetized AVM recursive verifier circuit; proof is {\pi_M, \pi_G}
+        InnerCircuitOutput inner_output =
+            construct_and_prove_inner_recursive_verification_circuit(stdlib_proof, public_inputs, input_agg_obj);
+
+        // Construct the outer Ultra-arithmetized Mega/Goblin recursive verifier circuit
+        RecursiveAvmGoblinOutput result =
+            construct_outer_recursive_verification_circuit(stdlib_proof, public_inputs, input_agg_obj, inner_output);
+
+        // Return ipa proof, ipa claim and output aggregation object produced from verifying the Mega + Goblin proofs
+        return result;
+    }
+
+    RecursiveAvmGoblinOutput construct_outer_recursive_verification_circuit(
+        const StdlibProof<UltraBuilder>& stdlib_proof,
+        const std::vector<std::vector<UltraFF>>& public_inputs,
+        AggregationObject input_agg_obj,
+        InnerCircuitOutput& inner_output) const
```

Comment:
Are these non const non references for a reason? (also in the other methods). I don't know the size but the could be being copied every single time.

Reply:
nope! added some consts and &s (everything here is indeed small)